src/boat_simulator/tests/unit/nodes/low_level_control/test_controller.py

```diff
@@ -0,0 +1,212 @@
+    # Test compute_error method of RudderController
+    error = rudder_controller._compute_error()
+    desired_rad = np.deg2rad(rudder_controller.desired_heading)
+    current_rad = np.deg2rad(rudder_controller.current_heading)
+    exp_error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+    assert np.equal(error, exp_error)
+
+
+def test_compute_setpoint(rudder_controller):
+    # Test compute_setpoint method of RudderController
+    setpoint = rudder_controller._compute_setpoint()
+    heading_error = rudder_controller._compute_error()  # heading_error in radians
+
+    rudder_setpoint = (rudder_controller.kp * heading_error) / (
+        1 + (rudder_controller.cp * abs(heading_error))
+    )
+
+    rudder_setpoint = min(
+        max(rudder_setpoint, np.deg2rad(rudder_controller.max_angle_range[0])),
+        np.deg2rad(rudder_controller.max_angle_range[1]),
+    )
+
+    rudder_setpoint_deg = np.rad2deg(rudder_setpoint)  # in degrees
+    running_error = rudder_setpoint_deg - rudder_controller.current_control_ang  # in degrees
+    assert np.equal(setpoint, rudder_setpoint_deg)
+    assert np.equal(running_error, rudder_controller.running_error)
+
+
+def test_reset_setpoint(rudder_controller):
+    # Test reset_setpoint method of RudderController
+    new_desired_heading = 60
+    new_current_heading = 30
+    expected = rudder_controller.reset_setpoint(new_desired_heading, new_current_heading)
+    rudder_controller._change_desired_heading(new_desired_heading)
+    calculated = rudder_controller._compute_setpoint()
+    np.equal(calculated, expected)
+
+
+# testing for one iteration only
+def test_update_state(rudder_controller):
+    current_control = rudder_controller.current_control_ang
+    rudder_controller._compute_setpoint()
+    assert rudder_controller.update_state() == (abs(rudder_controller.running_error) == 0)
+    if rudder_controller.current_heading != rudder_controller.desired_heading:
+        assert np.equal(
+            current_control
+            + (
+                copysign(
+                    rudder_controller.control_speed * rudder_controller.time_step,
+                    rudder_controller.running_error,
+                )
+            ),
+            rudder_controller.current_control_ang,
+        )
+
+
+def test_update_state_continuous(rudder_controller):
+    rudder_controller._compute_setpoint()
+    progress = False
+    counter = 0
+    if np.isclose(rudder_controller.running_error, 0, 0.1):
+        progress = True
+    else:
+        while not np.isclose(rudder_controller.running_error, 0, 0.1):
+            counter += 1
+            rudder_controller.update_state()
+            if rudder_controller.running_error == 0:
+                progress = True
+                break
+            if counter > 1000:
+                break
+    assert progress
+    assert np.equal(rudder_controller.current_control_ang, rudder_controller.setpoint)
```

Comment:
Check if they are close rather than for exact equality. In general, try not to compare for exact equality in floats cause you might get numerical issues.